core/network/src/protocol.rs

```diff
@@ -652,44 +859,55 @@ impl<B: BlockT, S: NetworkSpecialization<B>, H: ExHashT> Protocol<B, S, H> {
 	}
 
 	/// Called when we propagate ready extrinsics to peers.
-	pub fn propagate_extrinsics(&self, io: &mut SyncIo) {
+	fn propagate_extrinsics(&mut self) {
 		debug!(target: "sync", "Propagating extrinsics");
 
 		// Accept transactions only when fully synced
-		if self.sync.read().status().state != SyncState::Idle {
+		if self.sync.status().state != SyncState::Idle {
 			return;
 		}
 
 		let extrinsics = self.transaction_pool.transactions();
-
+		// TODO: see if self.send_message can be made &self,
+		// so that this vec would become uncessary.
+		let mut will_send = vec![];
 		let mut propagated_to = HashMap::new();
-		let mut peers = self.context_data.peers.write();
-		for (who, ref mut peer) in peers.iter_mut() {
+		for (who, ref mut peer) in self.context_data.peers.iter_mut() {
 			let (hashes, to_send): (Vec<_>, Vec<_>) = extrinsics
 				.iter()
 				.filter(|&(ref hash, _)| peer.known_extrinsics.insert(hash.clone()))
 				.cloned()
 				.unzip();
 
 			if !to_send.is_empty() {
-				let node_id = io.peer_id(*who).map(|id| id.to_base58());
+				let (sender, port) = unbounded();
+				let _ = self
+					.network_chan
+					.send(NetworkMsg::GetPeerId(who.clone(), sender));
+				let node_id = port.recv().expect("Failed to receive GetPeerId response");
```

Comment:
TBH i don't like blocking on waiting for messages here. The expectation is not really a convincing argument of impossibility either.

Reply:
I also don't like the blocking here, and I am less concerned about it than in the futures case discussed previously, because this only blocks the Protocol thread and nothing else. 

It would off-course be better to replace this with a non-blocking workflow, and I believe @tomaka has various plans to move some logic currently here into `network-libp2p`, what is going on here is perhaps a good candidate for that...

I have updated the expectation.